File: cli/internal_testsuites/golang/testsuite/exec_command_test/exec_command_test.go
package exec_command_test

import (
	"context"
	"github.com/kurtosis-tech/kurtosis-cli/golang_internal_testsuite/test_helpers"
	"github.com/kurtosis-tech/kurtosis-core-api-lib/api/golang/lib/services"
	"github.com/kurtosis-tech/stacktrace"
	"github.com/sirupsen/logrus"
	"github.com/stretchr/testify/require"
	"testing"
)

const (
	testName = "exec-command"
	isPartitioningEnabled = false

	execCmdTestImage      = "alpine:3.12.4"
	inputForLogOutputTest = "hello"
	expectedLogOutput     = "hello\n"
	testServiceId         = "test"

	successExitCode int32 = 0

)

var execCommandThatShouldWork = []string{
	"true",
}

var execCommandThatShouldHaveLogOutput = []string{
	"echo",
	inputForLogOutputTest,
}

var execCommandThatShouldFail = []string{
	"false",
}

func TestExecCommand(t *testing.T) {
	ctx := context.Background()

	// ------------------------------------- ENGINE SETUP ----------------------------------------------
	enclaveCtx, stopEnclaveFunc, _, err := test_helpers.CreateEnclave(t, ctx, testName, isPartitioningEnabled)
	require.NoError(t, err, "An error occurred creating an enclave")
	defer stopEnclaveFunc()

	// ------------------------------------- TEST SETUP ----------------------------------------------
	containerConfigSupplier := getContainerConfigSupplier()

	testServiceContext, err := enclaveCtx.AddService(testServiceId, containerConfigSupplier)
	require.NoError(t, err, "An error occurred starting service '%v'", testServiceId)

	// ------------------------------------- TEST RUN ----------------------------------------------
	logrus.Infof("Running exec command '%v' that should return a successful exit code...", execCommandThatShouldWork)
	shouldWorkExitCode, _, err := runExecCmd(testServiceContext, execCommandThatShouldWork)
	require.NoError(t, err, "An error occurred running exec command '%v'", execCommandThatShouldWork)
	require.Equal(t, successExitCode, shouldWorkExitCode, "Exec command '%v' should work, but got unsuccessful exit code %v", execCommandThatShouldWork, shouldWorkExitCode)
	logrus.Info("Exec command returned successful exit code as expected")

	logrus.Infof("Running exec command '%v' that should return an error exit code...", execCommandThatShouldFail)
	shouldFailExitCode, _, err := runExecCmd(testServiceContext, execCommandThatShouldFail)
	require.NoError(t, err, "An error occurred running exec command '%v'", execCommandThatShouldFail)
	require.NotEqual(t, successExitCode, shouldFailExitCode, "Exec command '%v' should fail, but got successful exit code %v", execCommandThatShouldFail, successExitCode)

	logrus.Infof("Running exec command '%v' that should return log output...", execCommandThatShouldHaveLogOutput)
	shouldHaveLogOutputExitCode, logOutput, err := runExecCmd(testServiceContext, execCommandThatShouldHaveLogOutput)
	require.NoError(t, err, "An error occurred running exec command '%v'", execCommandThatShouldHaveLogOutput)
	require.Equal(t, successExitCode, shouldHaveLogOutputExitCode, "Exec command '%v' should work, but got unsuccessful exit code %v", execCommandThatShouldHaveLogOutput, shouldHaveLogOutputExitCode)
	require.Equal(t, expectedLogOutput, logOutput, "Exec command '%v' should return %v, but got %v.", execCommandThatShouldHaveLogOutput, inputForLogOutputTest, logOutput)
	logrus.Info("Exec command returned error exit code as expected")
}

// ====================================================================================================
//                                       Private helper functions
// ====================================================================================================
func getContainerConfigSupplier() func(ipAddr string, sharedDirectory *services.SharedPath) (*services.ContainerConfig, error) {
	containerConfigSupplier  := func(ipAddr string, sharedDirectory *services.SharedPath) (*services.ContainerConfig, error) {

		// We sleep because the only function of this container is to test Docker executing a command while it's running
		// NOTE: We could just as easily combine this into a single array (rather than splitting between ENTRYPOINT and CMD
		// args), but this provides a nice little regression test of the ENTRYPOINT overriding
		entrypointArgs := []string{"sleep"}
		cmdArgs := []string{"30"}

		containerConfig := services.NewContainerConfigBuilder(
			execCmdTestImage,
		).WithEntrypointOverride(
			entrypointArgs,
		).WithCmdOverride(
			cmdArgs,
		).Build()
		return containerConfig, nil
	}
	return containerConfigSupplier
}

func runExecCmd(serviceContext *services.ServiceContext, command []string) (int32, string, error) {
	exitCode, logOutput, err := serviceContext.ExecCommand(command)
	if err != nil {
		return 0, "", stacktrace.Propagate(
			err,
			"An error occurred executing command '%v'", command)
	}
	return exitCode, logOutput, nil
}

File: cli/internal_testsuites/typescript/src/testsuite/exec_command_test/exec_command.test.ts
import { ContainerConfig, ContainerConfigBuilder, ServiceContext } from "kurtosis-core-api-lib";
import log from "loglevel";
import { err, ok, Result } from "neverthrow";

import { createEnclave } from "../../test_helpers/enclave_setup";

const TEST_NAME = "exec-command";
const IS_PARTITIONING_ENABLED = false;
const EXEC_CMD_TEST_IMAGE = "alpine:3.12.4";
const INPUT_FOR_LOG_OUTPUT_TEST = "hello";
const EXPECTED_LOG_OUTPUT = "hello\n";
const INPUT_FOR_ADVANCED_LOG_OUTPUT_TEST = "hello && hello";
const EXPECTED_ADVANCED_LOG_OUTPUT = "hello && hello\n"
const TEST_SERVICE_ID = "test";
const SUCCESS_EXIT_CODE = 0;

const EXEC_COMMAND_THAT_SHOULD_WORK = ["true"]
const EXEC_COMMAND_THAT_SHOULD_HAVE_LOG_OUTPUT = ["echo", INPUT_FOR_LOG_OUTPUT_TEST]
// This command tests to ensure that the commands the user is running get passed exactly as-is to the Docker
// container. If Kurtosis code is magically wrapping the code with "sh -c", this will fail.
const EXEC_COMMAND_THAT_WILL_FAIL_IF_SH_WRAPPED = [
    "echo",
    INPUT_FOR_ADVANCED_LOG_OUTPUT_TEST,
]
const EXEC_COMMAND_THAT_SHOULD_FAIL = ["false"]

jest.setTimeout(180000)

test("Test exec command", async () => {
    // ------------------------------------- ENGINE SETUP ----------------------------------------------
    const createEnclaveResult = await createEnclave(TEST_NAME, IS_PARTITIONING_ENABLED)

    if(createEnclaveResult.isErr()) { throw createEnclaveResult.error }
    
    const { enclaveContext, stopEnclaveFunction } = createEnclaveResult.value

    try {
        // ------------------------------------- TEST SETUP ----------------------------------------------
        const containerConfig = getContainerConfig()

        const addServiceResult = await enclaveContext.addService(TEST_SERVICE_ID, containerConfig)

        if(addServiceResult.isErr()) {
            log.error(`An error occurred starting service "${TEST_SERVICE_ID}"`);
            throw addServiceResult.error
        };

        const testServiceContext = addServiceResult.value

        // ------------------------------------- TEST RUN ----------------------------------------------
        log.info(`Running exec command "${EXEC_COMMAND_THAT_SHOULD_WORK}" that should return a successful exit code...`)
        
        const runExecCmdShouldWorkResult = await runExecCmd(testServiceContext, EXEC_COMMAND_THAT_SHOULD_WORK)

        if(runExecCmdShouldWorkResult.isErr()){
            log.error(`An error occurred running exec command "${EXEC_COMMAND_THAT_SHOULD_WORK}"`)
            throw runExecCmdShouldWorkResult.error
        }
        const [ shouldWorkExitCode ] = runExecCmdShouldWorkResult.value;

        if(SUCCESS_EXIT_CODE !== shouldWorkExitCode){
            throw new Error(`Exec command "${EXEC_COMMAND_THAT_SHOULD_WORK}" should work, but got unsuccessful exit code ${shouldWorkExitCode}`)
        }

        log.info("Exec command returned successful exit code as expected")

        log.info(`Running exec command "${EXEC_COMMAND_THAT_SHOULD_FAIL}" that should return an error exit code...`)

        const runExecCmdShouldFailResult = await runExecCmd(testServiceContext, EXEC_COMMAND_THAT_SHOULD_FAIL)

        if(runExecCmdShouldFailResult.isErr()){
            log.error(`An error occurred running exec command "${EXEC_COMMAND_THAT_SHOULD_FAIL}"`)
            throw runExecCmdShouldFailResult.error
        }
        const [ shouldFailExitCode ] = runExecCmdShouldFailResult.value;

        if(SUCCESS_EXIT_CODE === shouldFailExitCode){
            throw new Error(`Exec command "${EXEC_COMMAND_THAT_SHOULD_FAIL}" should fail, but got successful exit code ${SUCCESS_EXIT_CODE}`)
        }
        log.info("Exec command returning an error exited with error");

        log.info(`Running exec command '${EXEC_COMMAND_THAT_SHOULD_HAVE_LOG_OUTPUT}' that should return log output...`)
        const runExecCmdShouldLogOutputResult = await runExecCmd(testServiceContext, EXEC_COMMAND_THAT_SHOULD_HAVE_LOG_OUTPUT)
        if(runExecCmdShouldLogOutputResult.isErr()){
            log.error(`An error occurred running exec command '${EXEC_COMMAND_THAT_SHOULD_HAVE_LOG_OUTPUT}'`)
            throw runExecCmdShouldLogOutputResult.error
        }
        const [ shouldHaveLogOutputExitCode, logOutput ] = runExecCmdShouldLogOutputResult.value;
        if(SUCCESS_EXIT_CODE !== shouldHaveLogOutputExitCode){
            throw new Error(`Exec command '${EXEC_COMMAND_THAT_SHOULD_HAVE_LOG_OUTPUT}' should work, but got unsuccessful exit code ${shouldHaveLogOutputExitCode}`)
        }
        if(EXPECTED_LOG_OUTPUT !== logOutput){
            throw new Error(`Exec command '${EXEC_COMMAND_THAT_SHOULD_HAVE_LOG_OUTPUT}' should return '${EXPECTED_LOG_OUTPUT}', but got ${logOutput}`)
        }
        log.info("Exec command returning log output passed as expected");

        log.info(`Running exec command '${EXEC_COMMAND_THAT_WILL_FAIL_IF_SH_WRAPPED}' that will fail if Kurtosis is accidentally sh-wrapping the command...`)
        const shouldFailIfShWrappedExecCmdResult = await runExecCmd(testServiceContext, EXEC_COMMAND_THAT_WILL_FAIL_IF_SH_WRAPPED)
        if(shouldFailIfShWrappedExecCmdResult.isErr()){
            log.error(`An error occurred running exec command '${EXEC_COMMAND_THAT_WILL_FAIL_IF_SH_WRAPPED}'`)
            throw shouldFailIfShWrappedExecCmdResult.error
        }
        const [ shouldNotGetShWrappedExitCode, shouldNotGetShWrappedLogOutput ] = shouldFailIfShWrappedExecCmdResult.value;
        if(SUCCESS_EXIT_CODE !== shouldNotGetShWrappedExitCode){
            throw new Error(`Exec command '${EXEC_COMMAND_THAT_WILL_FAIL_IF_SH_WRAPPED}' should work, but got unsuccessful exit code ${shouldNotGetShWrappedExitCode}`)
        }
        if(EXPECTED_ADVANCED_LOG_OUTPUT !== shouldNotGetShWrappedLogOutput){
            throw new Error(`Exec command '${EXEC_COMMAND_THAT_WILL_FAIL_IF_SH_WRAPPED}' should return '${EXPECTED_ADVANCED_LOG_OUTPUT}', but got ${shouldNotGetShWrappedLogOutput}`)
        }
        log.info("Exec command that will fail if Kurtosis is accidentally sh-wrapping did not fail");

    }finally{
        stopEnclaveFunction()
    }
})

// ====================================================================================================
//                                       Private helper functions
// ====================================================================================================
function getContainerConfig(): ContainerConfig {
    const entrypointArgs = ["sleep"]
    const cmdArgs = ["30"]

    const containerConfig = new ContainerConfigBuilder(EXEC_CMD_TEST_IMAGE)
        .withEntrypointOverride(entrypointArgs)
        .withCmdOverride(cmdArgs)
        .build()

    return containerConfig
}

async function runExecCmd(serviceContext: ServiceContext, command: string[]) {
    const execCommandResult = await serviceContext.execCommand(command)
    if(execCommandResult.isErr()) {
        return err(execCommandResult.error)
    }
    return ok(execCommandResult.value)
}